Merge Bernhard's theme live preview, add README (#745)
digital-asset/docs.daml.com: theme/docs/templates.rst

          debug desc

Turning off the warning
=======================

This warning is controlled by the warning flag ``template-let``, which means
that it can be toggled independently of other warnings. This is especially
useful for gradually migrating code that used this syntax.

To turn off the warning within a Daml file, add the following line at the top of
the file:

.. code-block:: daml

  {-# OPTIONS_GHC -Wno-template-let #-}

To turn it off for an entire Daml project, add the following entry to the
``build-options`` field of the project's ``daml.yaml`` file

.. code-block:: yaml

  build-options:
  - --ghc-option=-Wno-template-let

Within a project where the warning has been turned off via the ``daml.yaml``
file, it can be turned back on for individual Daml files by adding the following
line at the top of each file:

.. code-block:: daml

  {-# OPTIONS_GHC -Wtemplate-let #-}

.. _daml-ref-signatories:

Signatory Parties
*****************

.. literalinclude:: ../code-snippets/Reference.daml
   :language: daml
   :start-after: -- start template sigs snippet
   :end-before: -- end template sigs snippet

> - ``signatory`` keyword. After ``where``. Followed by at least one ``Party``.

Reviewer: ```suggestion
- Signatories are defined with the ``signatory`` keyword , which takes one or more ``Party``s as arguments, in the ``where`` block.
```

  Daml won't let you put someone into an obligable position without their consent. So if the contract will cause obligations for a party, they *must* be a signatory. **If they haven't authorized it, you won't be able to create the contract.** In this situation, you may see errors like:

  ``NameOfTemplate requires authorizers Party1,Party2,Party, but only Party1 were given.``
- When a signatory consents to the contract creation, this means they also authorize the consequences of :ref:`choices <daml-ref-choices>` that can be exercised on this contract.
- The contract is visible to all signatories (as well as the other stakeholders of the contract). That is, the compiler automatically adds signatories as observers.
- Each template **must** have at least one signatory. A signatory declaration consists of the `signatory` keyword followed by a comma-separated list of one or more expressions, each expression denoting a ``Party`` or collection thereof.

.. _daml-ref-observers:

Observers
*********

.. literalinclude:: ../code-snippets/Reference.daml
   :language: daml
   :start-after: -- start template obs snippet
   :end-before: -- end template obs snippet

> - ``observer`` keyword. After ``where``. Followed by at least one ``Party``.

Reviewer: ```suggestion
- Observers are defined with the ``observer`` keyword , which takes one or more ``Party``s as arguments, in the ``where`` block.
```

- Observers are additional stakeholders, so the contract is visible to these parties (see the ``Party`` type).
- Optional. You can have many, either as a comma-separated list or reusing the keyword. You could pass in a list (of type ``[Party]``).
- Use when a party needs visibility on a contract, or be informed or contract events, but is not a :ref:`signatory <daml-ref-signatories>` or :ref:`controller <daml-ref-controllers>`.
- If you start your choice with ``choice`` rather than ``controller`` (see :ref:`daml-ref-choices` below), you must make sure to add any potential controller as an observer. Otherwise, they will not be able to exercise the choice, because they won't be able to see the contract.

.. _daml-ref-choices:

Choices
*******

.. literalinclude:: ../code-snippets/Reference.daml
   :language: daml
   :start-after: -- start template choice snippet
   :end-before: -- end template choice snippet

> - A right that the contract gives the controlling party. Can be *exercised*.

Reviewer: ```suggestion
- A choice is an action that a controller party can *exercise* on a contract.
```

- This is essentially where all the logic of the template goes.

> - By default, choices are *consuming*: that is, exercising the choice archives the contract, so no further choices can be exercised on it. You can make a choice non-consuming using the ``nonconsuming`` keyword.

Reviewer: ```suggestion
- By default, choices are *consuming*. Exercising a choice archives the contract and no further choices can be exercised on it. You can make a choice non-consuming using the ``nonconsuming`` keyword.
```

- See :doc:`choices` for full reference information.

.. _daml-ref-serializable-types:

Serializable Types
******************

Every parameter to a template, choice argument, and choice result must have a *serializable type*.
This does not merely mean "convertible to bytes"; it has a specific meaning in Daml.
The serializability rule serves three purposes:

1. Offer a stable means to store ledger values permanently.
2. Provide a sensible encoding of them over :doc:`/app-dev/ledger-api`.
3. Provide sensible *types* that directly match their Daml counterparts in languages like Java for language codegen.

For example, certain kinds of type parameters Daml offers are compatible with (1) and (2), but have no proper counterpart in (3), so they are disallowed.
Similarly, function types have sensible Java counterparts, satisfying (3), but no reliable way to store or share them via the API, thus failing (1) and (2).

> The following types are *not serializable*, and thus may not be used in templates.

Reviewer: ```suggestion
The following types are *not serializable*, and may not be used in templates.
```

- Function types.
- Record types with any non-serializable field.
- Variant types with any non-serializable value case.
- Variant and enum types with no constructors.
- References to a parameterized data type with any non-serializable type argument.
  This applies whether or not the data type definition uses the type parameter.
- Defined data types with any type parameter of kind ``Nat``, or any kind other than ``*``.
  This means higher-kinded types, and types that take a parameter just to pass to ``Numeric``, are not serializable.

.. _daml-ref-agreements:

Agreements (Deprecated)
***********************

.. literalinclude:: ../code-snippets/Reference.daml
   :language: daml
   :start-after: -- start template agree snippet
   :end-before: -- end template agree snippet

> - ``agreement`` keyword, followed by text.

Reviewer: ```suggestion
- Agreements are defined with the ``agreement`` keyword, followed by text.
```

> - Represents what the contract means in text. They're usually the boundary between on-ledger and off-ledger rights and obligations.

Reviewer: ```suggestion
- Agreements represent what the contract means in text. They're usually the boundary between on-ledger and off-ledger rights and obligations.
```

- Usually, they look like ``agreement tx``, where ``tx`` is of type ``Text``.

  You can use the built-in operator ``show`` to convert party names to a string, and concatenate with ``<>`` .

.. warning::
   Since Daml 2.9.0, agreement declarations are deprecated, and their
   presence will result in the following warning:

   .. code-block:: text

      Found 'agreement' declaration in template definition.
      This feature is deprecated, it will be removed in a future version of Daml.

Migration
=========

Users are encouraged to remove any agreement declarations from their code, to

> ensure it will continue working once this feature is fully removed from the language.

Reviewer: ```suggestion
ensure that it continues to work once this feature is fully removed from the language.
```

Turning off the warning
=======================

This warning is controlled by the warning flag ``template-agreement``, which means
that it can be toggled independently of other warnings. This is especially useful
for gradually migrating code that used this feature.

To turn off the warning within a Daml file, add the following line at the top of
the file:

.. code-block:: daml

  {-# OPTIONS_GHC -Wno-template-agreement #-}

To turn it off for an entire Daml project, add the following entry to the
``build-options`` field of the project's ``daml.yaml`` file:

.. code-block:: yaml

  build-options:
  - --ghc-option=-Wno-template-agreement

Within a project where the warning has been turned off via the ``daml.yaml``
file, it can be turned back on for individual Daml files by adding the following
line at the top of each file:

.. code-block:: daml

  {-# OPTIONS_GHC -Wtemplate-agreement #-}

.. _daml-ref-preconditions:

Preconditions
*************

.. literalinclude:: ../code-snippets/Reference.daml
   :language: daml
   :start-after: -- start template ensure snippet
   :end-before: -- end template ensure snippet

> - ``ensure`` keyword, followed by a boolean condition.

Reviewer: ```suggestion
- Preconditions are defined with the ``ensure`` keyword, followed by a boolean condition.
```

- Used on contract creation. ``ensure`` limits the values on parameters that can be passed to the contract: the contract can only be created if the boolean condition is true.

.. _daml-ref-contract-keys:

.. _daml-ref-maintainers:

Contract Keys and Maintainers
*****************************

.. literalinclude:: ../code-snippets/Reference.daml
   :language: daml
   :start-after: -- start contract key snippet
   :end-before: -- end contract key snippet

> - ``key`` and ``maintainer`` keywords.

Reviewer: ```suggestion
- Contract keys and maintainers are defined with the ``key`` and ``maintainer`` keywords.
```

> - This feature lets you specify a "key" that you can use to uniquely identify this contract as an instance of this template.

Reviewer: ```suggestion
- This feature lets you specify a "key" that you can use to uniquely identify a contract as an instance of a template.
```

- If you specify a ``key``, you must also specify a ``maintainer``. This is a ``Party`` that will ensure the uniqueness of all the keys it is aware of.

  Because of this, the ``key`` must include the ``maintainer`` ``Party`` or parties (for example, as part of a tuple or record), and the ``maintainer`` must be a signatory.
- For a full explanation, see :doc:`/daml/reference/contract-keys`.

Interface Instances
*******************

.. literalinclude:: ../code-snippets-dev/Interfaces.daml
   :language: daml
   :start-after: -- INTERFACE_INSTANCE_IN_TEMPLATE_BEGIN
   :end-before: -- INTERFACE_INSTANCE_IN_TEMPLATE_END

> - Used to make a template an instance of an existing interface.

Reviewer: ```suggestion
- An interface instance is used to make a template an instance of an existing interface.
```
